docs: add september showcase post (#30)
e18e/e18e: docs/blog/september-contributions-showcase.md

      content: September contributions showcase
  - - meta
    - property: og:url
      content: https://e18e.dev/blog/september-contributions-showcase
  - - meta
    - property: og:description
      content: A showcase of e18e-focused contributions from September
---

# September contributions

_October 3, 2024_

It's that time of month again where we show off some of the great contributions we've seen around the e18e community in the past month!

# Libraries

This month has been a great one for improvements to existing and upcoming libraries. Many have been listening closely to the e18e community and pushing forward together.

## chokidar 4.x

[chokidar](https://github.com/paulmillr/chokidar) is a library for watching for changes on a file system. Built on top of node's own watcher capabilities, it provides a higher level and simpler interface.

Almost all of us depend on this library in some way, often through the toolchain or framework we use.

Paul ([@paulmillr](https://x.com/paulmillr/), the author of chokidar) and I ([@43081j](https://x.com/43081j)) have been working hard to get the next major version of it over the line for a long time now (the branch was made in 2021!).

> Recently, we finally released [4.0.0](https://github.com/paulmillr/chokidar/releases/tag/4.0.0)! :partying_face: this massively reduces the size and complexity.

Reviewer: ```suggestion
Recently, we finally released [4.0.0](https://github.com/paulmillr/chokidar/releases/tag/4.0.0)! :partying_face: This massively reduces the size and complexity.
```

Probably a typo.

Those of you who remember the `is-number` fiasco on social media will now also be super happy that we (chokidar) were the largest consumer of it and have since dropped it entirely :tada:

Big wins for the e18e community!

Thanks to [@benjaminmccann](https://x.com/benjaminmccann) and talentlessguy for contributing _many_ v3 to v4 upgrades already!

## fdir / tinyglobby

`fdir` was in last month's showcase but has earned another mention thanks to the author ([thecodrr](https://x.com/thecodrr)) landing various community contributed PRs and features.

This is a great example of where an author has joined the community and worked closely with it to further their library, and provide some much needed fixes and features.

In version [6.4.0](https://github.com/thecodrr/fdir/releases/tag/v6.4.0), we saw two important changes:

- Ability to exclude symlinks
- Ability to bring your own glob library (e.g. [zeptomatch](https://github.com/fabiospampinato/zeptomatch))

These two changes have helped the community migrate even more high level tools and frameworks across.

> Similarly, tinyglobby has shipped a few new versions. Particularly it is now able to determine if a string is a glob or not. This is super useful for migrating more projects to it.

Reviewer 2: ```suggestion
Similarly, [tinyglobby](https://github.com/SuperchupuDev/tinyglobby) has shipped a few new versions. Particularly it is now able to determine if a string is a glob or not. This is super useful for migrating more projects to it.
```

> Thanks to pralkarz, [@benjaminmccann](https://x.com/benjaminmccann) and [@superchupudev](https://x.com/superchupu) for already migrating so many projects to these two libraries!

Reviewer: ```suggestion
Thanks to [pralkarz](https://github.com/ziebam), [@benjaminmccann](https://x.com/benjaminmccann) and [@superchupudev](https://x.com/superchupu) for already migrating so many projects to these two libraries!
```

Could you link my GitHub profile there?

## tinyexec migrations

> The community has been working wonders migrating countless projects from `execa` and other heavyweight libraries to the super lean [tinyexec](https://github.com/tinylibs/tinyexec/).

Reviewer 2: Should we try to avoid calling other libraries heavyweight to avoid any sensitivity there?
```suggestion
The community has been working wonders migrating countless projects from `execa`, `ez-spawn`, and other libraries to the super lean [tinyexec](https://github.com/tinylibs/tinyexec/).
```

Author: You're right, I mostly meant that are "full fat" (in that they have all the extra features you might not usually need) 

But I prefer your wording 👍

> For example, thanks to pralkarz, [vitest](https://github.com/vitest-dev/vitest) has already moved over!

Reviewer: ```suggestion
For example, thanks to [pralkarz](https://github.com/ziebam), [vitest](https://github.com/vitest-dev/vitest) has already moved over!
```

Same here. 🙂

There are many more PRs currently in progress to do just the same :pray: Great work by the community here.

# Contributions / Improvements

> ## storybook
> 
> The [storybook](https://storybook.js.org/) team have been working closely with the e18e community for some time now, dedicating huge amounts of time to improving the performance and install footprint for everyone.

Reviewer 3: ```suggestion
## Storybook

The [Storybook](https://storybook.js.org/) team have been working closely with the e18e community for some time now, dedicating huge amounts of time to improving the performance and install footprint for everyone.
```

This month, they've kept up the pace! Just a few examples of what has landed:

- Migrated from `chalk` to `picocolors`

> - Migrated from `fs-extra` to built in `fs` functions

Reviewer 2: Should mention ziebam somewhere as it was a big commit. https://github.com/storybookjs/storybook/pull/29126

Reviewer 3: ```suggestion
- [pralkarz](https://github.com/ziebam) migrated from `fs-extra` to built in `fs` functions
```

- Removed `handlebars` and simplified templating internally

> - Removed `prettier` from the core bundle (saved **4MB** on the install size)

Reviewer 3: I double-checked the commits, this actually shaved off 10 MB because we're dual-publishing the bundle
```suggestion
- Removed `prettier` from the core bundle (saved **10MB** on the install size)
```

- Replaced `lodash` with `es-toolkit` (more huge savings, especially with tree shaking)

On top of this, the team has actually started a full project focusing on [reducing the install footprint](https://github.com/storybookjs/storybook/issues/29038) in collaboration with the e18e community.

For anyone interested in helping out in the e18e space, this issue a great way to learn some of the techniques used to track down bundle/install size issues.

> As a bonus, they even integrated bundle size analysis into their CI so all new PRs can be compared with the main branch. Would love to see that in more projects!

Reviewer 2: Could be cool to add a link to where this was setup or some info about it to help others who might want to do the same

Reviewer 3: I don't know if this is _too_ detailed:

```suggestion
As a bonus, they even integrated visual bundle size analysis into their CI so all new PRs can be compared with the main branch. They did this by:

1. Generating [esbuild metafiles](https://esbuild.github.io/api/#build-metadata) for all of their packages every time they're built
2. Importing those metafiles in a story in the repository's internal Storybook.
3. That story sneakily passes the chosen metafile into the [esbuild Bundle Size Analyzer website](https://esbuild.github.io/analyze/) via a query parameter, which then gives a nice visual overview of the bundled output.

You can [play around with the result in their Storybook](https://next--635781f3500dd2c49e189caf.chromatic.com/?path=/story/bench--es-build-analyzer&args=metafile:builder-vite__metafile) and navigate through the different packages via the Controls panel at the bottom.

Would love to see that in more projects! You can get inspiration from [the PR](https://github.com/storybookjs/storybook/pull/29117) that implemented the work. Storybook isn't necessary for any of this to work (it was just a convenient entry point for them) - all of this can be achieved with cleverly crafted URLs and some patience. ✨
```

Happy to reword it if necessary, to make it more approachable for the non-Storybook audience.

Author: I think it might be. It almost justifies it's own blog post from your end I think 👀

If you could think of a good way to make it a little more of a summary, that would be super helpful (as a suggestion again so I can commit from the GitHub UI 🙏)

Reviewer 3: Is this better?

```suggestion
As a bonus, they even integrated visual bundle size analysis into their CI so all new PRs can be compared with the main branch.
They did this by generating [esbuild metafiles](https://esbuild.github.io/api/#build-metadata) for all of their packages every time they're built and storing links to the [esbuild Bundle Size Analyzer website](https://esbuild.github.io/analyze/) with those metafiles embedded.
The final result is a visual overview of all package's bundle, that you can [play around with here](https://next--635781f3500dd2c49e189caf.chromatic.com/?path=/story/bench--es-build-analyzer&args=metafile:builder-vite__metafile).

Would love to see that in more projects! You can get inspiration from [the PR](https://github.com/storybookjs/storybook/pull/29117) that implemented the work.
```

Author: Looks good to me! 🥳

This has all been so great to see and will hopefully lead to other large projects making similar efforts! Definitely some big wins ahead :raised_hands:

## nx

The folks over at [nx](https://github.com/nrwl/nx/) have been incredibly active with the e18e community. They are definitely on the same page and have been moving fast to work with the community to land many improvements.

You can see a lot of this in the [tracking issue](https://github.com/es-tooling/ecosystem-cleanup/issues/117) over on the ecosystem cleanup project. These issues are closing quickly, with contributions coming from a few different people in the e18e space.

Just a few examples of the improvements made:

- Migrating older fs-extra code to use newer, native `fs` functionality
- Migrating from chalk to picocolors
- Migrating from glob libraries to fdir

These are all improving the CPU performance of nx and cutting huge subtrees out of the dependency tree at the same time.

## jimp

> [jimp](https://github.com/jimp-dev/jimp), an image manipulation library, has dropped an impressive amount of dependencies and reduced their package size hugely.

Reviewer 2: This was kind of a wash. They introduced a new version that was a big regression. I noticed it while upgrading and it was a pretty small change to fix it

Reviewer 2: It was jimp #1302. Basically I just removed rspack as an unused dependency plus a few other minor cleanups

Author: We should probably rewrite this to instead just mention the size reduction I think then

If I understood correctly 👀
